pkg/plugins/document/dynamodb/dynamodb.go

```diff
@@ -261,6 +262,42 @@ func (s *DynamoDocService) Delete(key *document.Key) error {
 	return nil
 }
 
+func (s *DynamoDocService) query(collection *document.Collection, expressions []document.QueryExpression, limit int, pagingToken map[string]string) (*document.QueryResult, error) {
+	queryResult := &document.QueryResult{
+		Documents: make([]document.Document, 0),
+	}
+
+	var resFunc resultRetriever = s.performQuery
+	if collection.Parent == nil || collection.Parent.Id == "" {
+		resFunc = s.performScan
+	}
+
+	if res, err := resFunc(collection, expressions, limit, pagingToken); err != nil {
+		return nil, err
+	} else {
+		queryResult.Documents = append(queryResult.Documents, res.Documents...)
+		queryResult.PagingToken = res.PagingToken
+	}
+
+	remainingLimit := limit - len(queryResult.Documents)
+
+	// If more results available, perform additional queries
+	for remainingLimit > 0 &&
+		(queryResult.PagingToken != nil && len(queryResult.PagingToken) > 0) {
+
+		if res, err := resFunc(collection, expressions, remainingLimit, queryResult.PagingToken); err != nil {
+			return nil, err
+		} else {
+			queryResult.Documents = append(queryResult.Documents, res.Documents...)
+			queryResult.PagingToken = res.PagingToken
+		}
+
+		remainingLimit = limit - len(queryResult.Documents)
+	}
```

Comment:
This is the eager code I mention in the next comment

Reply:
done